From d8b4dc808cbb6d2f7b037f23cad4edf61f8adefc Mon Sep 17 00:00:00 2001
From: Moritz Barsnick <moritz.barsnick@chargebyte.com>
Date: Tue, 13 Aug 2024 15:47:33 +0200
Subject: [PATCH] ChargePoint: add support for SecurityEventNotification
 `FirmwareUpdated` after reboot

It was being reported as ResetOrReboot.

This addition is to fulfil requirement L01.FR.31 when the firmware requires
a reboot after installation. The case without reboot is handled in
on_firmware_update_status_notification().

Signed-off-by: Moritz Barsnick <moritz.barsnick@chargebyte.com>
---
 lib/ocpp/v201/charge_point.cpp | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/lib/ocpp/v201/charge_point.cpp b/lib/ocpp/v201/charge_point.cpp
index 3c9ec43a..d63a4860 100644
--- a/lib/ocpp/v201/charge_point.cpp
+++ b/lib/ocpp/v201/charge_point.cpp
@@ -234,6 +234,12 @@ void ChargePoint::start(BootReasonEnum bootreason) {
             this->device_model->get_value<std::string>(ControllerComponentVariables::FirmwareVersion));
         this->security_event_notification_req(CiString<50>(ocpp::security_events::STARTUP_OF_THE_DEVICE),
                                               std::optional<CiString<255>>(startup_message), true, true);
+    } else if (this->bootreason == BootReasonEnum::FirmwareUpdate) {
+        std::string startup_message = "Charging station reboot after firmware update. Firmware version: ";
+        startup_message.append(
+            this->device_model->get_value<std::string>(ControllerComponentVariables::FirmwareVersion));
+        this->security_event_notification_req(CiString<50>(ocpp::security_events::FIRMWARE_UPDATED),
+                                              std::optional<CiString<255>>(startup_message), true, true);
     } else {
         std::string startup_message = "Charging station reset or reboot. Firmware version: ";
         startup_message.append(
-- 
2.34.1

